src/derparams.f90

```diff
@@ -1,10 +1,667 @@
+         call tdsops%interpl_mid(variation, from_to, bc_top, bc_bot, paire)
+      else if (operation == 'staggared-deriv') then
+         call tdsops%stag_der_1st(variation, from_to, bc_top, bc_bot, paire)
+      else
+         error stop 'operation is not defined'
+      end if
+
+      !if ((bc_top == 'periodic' .and. bc_bot /= 'periodic') .or. &
+      !    (bc_top /= 'periodic' .and. bc_bot == 'periodic')) then
+      !   error stop 'In case of periodic BCs both ends must be periodic'
+      !end if
+
+   end function tdsops
+
+   subroutine deriv_1st(self, variation, bc_top, bc_bot, paire)
+      implicit none
+
+      class(tdsops_t), intent(inout) :: self
+      character(*), intent(in) :: variation
+      character(*), optional, intent(in) :: bc_top, bc_bot
+      logical, optional, intent(in) :: paire
+
+      real(dp), allocatable :: dist_b(:)
+      real(dp) :: afi, bfi, dx
+      integer :: i, n, n_halo
+
+      dx = self%delta
+
+      ! alpha is alfa
+
+      select case(variation)
+      case('compact6')
+         self%alpha = 1._dp/3._dp
+         afi = 7._dp/9._dp/self%delta
+         bfi = 1._dp/36._dp/self%delta
+      case default
+         error stop 'variation is not defined'
+      end select
+
+      self%coeffs(:) = [0._dp, 0._dp, -bfi, -afi, &
+                        0._dp, &
+                        afi, bfi, 0._dp, 0._dp]
+
+      do i = 1, self%n_halo
+         self%coeffs_t(i,:) = self%coeffs(:)
+         self%coeffs_b(i,:) = self%coeffs(:)
+      end do
+
+      self%dist_sa(:) = self%alpha; self%dist_sc(:) = self%alpha
+
+      n = self%n
+      n_halo = self%n_halo
+
+      allocate(dist_b(n))
+      dist_b(:) = 1._dp
+
+      select case (bc_top)
+      case ('neumann')
+         if (present(paire) .and. paire) then
+            ! paire == 1; d(uu)/dx, dv/dx, dw/dx
+            !             d(vv)/dy, du/dy, dw/dy
+            !             d(ww)/dz, du/dz, dv/dz
+            self%dist_sa(1) = 0._dp
+            self%dist_sc(1) = 0._dp
+            self%coeffs_t(1, :) = [0._dp, 0._dp, 0._dp, 0._dp, &
+                                   0._dp, &
+                                   0._dp, 0._dp, 0._dp, 0._dp]
+            self%coeffs_t(2, :) = [0._dp, 0._dp, 0._dp, -afi, &
+                                   -bfi, &
+                                   afi, bfi, 0._dp, 0._dp]
+         else
+            ! paire == 0; d(uv)/dx, d(uw)/dx, du/dx
+            !             d(vu)/dy, d(vw)/dy, dv/dy
+            !             d(wu)/dz, d(wv)/dz, dw/dz
+            self%dist_sa(1) = 0._dp
+            self%dist_sc(1) = 2*self%alpha
+            self%coeffs_t(1, :) = [0._dp, 0._dp, 0._dp, 0._dp, &
+                                   0._dp, &
+                                   2*afi, 2*bfi, 0._dp, 0._dp]
+            self%coeffs_t(2, :) = [0._dp, 0._dp, 0._dp, -afi, &
+                                   bfi, &
+                                   afi, bfi, 0._dp, 0._dp]
+         end if
```

Comment:
Would it be neater to write these as helper subroutines? This subroutine would then be a "driver" and the utility subroutines easier to understand (and test) in isolation

Reply:
There is a similarity for the operator matrices accross diffrent types of operations, but we need to edit only a few entries based on the BC and the exact specification may differ in similar cases (for example der_1st, neumann and paire=0 is similar to der_2nd, neumann, paire=1, not paire=0). We can call this helper function to set things as we want but then it is already a single or a few line assignment, so I'm not sure delegating this simple assignment to a separate function.

For the RHS builder matrices it is almost entirely different for the first derivative, second derivative, interpolation, and staggared derivative. Only similarity is between the top and bottom BCs of the same operation if the BCs are the same. So there is a potential simplification here, but then the helper function needs to be a tad more elaborate to assign coefficients considering whether its a top or bottom, so not sure if it will be easier to understand at the end, haven't tried. Can you provide a very simple demonstration here if you have something in mind?

Reply:
I had in mind ```call set_neuman(..., paire); call set_dirichlet(...)``` etc. The advantage of this is then you can have automated tests that call these directly with a preprepared input. Ultimately this should be easier to debug than calling the "driver" function and checking its output to find where it went wrong (still a worthwhile test, the finer grained ones will help narrow down where exactly the error is).